### PR TITLE
Update ranges.hpp for v1/3

### PR DESCRIPTION
**Pull request by james-d-mitchell**
Should be good to go!

### CODE REVIEW

File: include/libsemigroups/ranges.hpp [hunk 1]
```diff
@@ -38,28 +38,101 @@
...
+    Random& operator=(Random&&) = default;
+
     template <typename InputRange>
     struct Range {
       using output_type = typename InputRange::output_type;
 
-      static constexpr bool is_finite     = true;
-      static constexpr bool is_idempotent = false;
+      // TODO(1) this is probably not correct, should depend on InputRange
+      static constexpr bool is_finite = true;
```
Comment (Joseph-Edwards) [inline include/libsemigroups/ranges.hpp:125]:
```suggestion
      static constexpr bool is_finite = rx::is_finite_v<InputRange>;
```
Something like this?

**Review by Joseph-Edwards — COMMENTED** [review]
All looks good to me! I've left a speculative suggestion about one of the TODOs below, but if that's not a good suggestion then this seems good to merge

File: include/libsemigroups/ranges.hpp [hunk 2]
```diff
@@ -38,28 +38,101 @@
...
+      // TODO(1) this is probably not correct, should depend on InputRange
+      static constexpr bool is_idempotent = true;
```
Comment (james-d-mitchell) [inline include/libsemigroups/ranges.hpp:127]:
```suggestion
      static constexpr bool is_idempotent = rx::is_idempotent_v<InputRange>;
```

**Comment by james-d-mitchell** [comment]
Let's see if that works :)

**Pull request merged**